### PR TITLE
FEAT: Improve antenna array processing and plotting

### PR DESCRIPTION
**Pull request by Devin-Crawford**
Minor improvements to plotting and functionality to support array phase excitations for beam-steering in HFSS.

### CODE REVIEW

File: pyaedt/application/analysis_hf.py [hunk 1]
```diff
@@ -360,3 +360,64 @@ def export_touchstone(
...
+    m : int, required
+        Index of the rectangular antenna array element in the x-direction.
+    n : int, required
+        Index of the rectangular antenna array element in the y-direction.
+    theta_name : str, optional
+        Post-processing variable name in HFSS that is used for the
+        theta component of the phase angle expression. Default is ``"theta_scan"``.
+    phi_name : str, optional
+        Post-processing variable name in HFSS to be used to generate
+        the phi component of the phase angle expression. Default is ``"phi_scan"``
```
Comment (PipKat) [inline pyaedt/application/analysis_hf.py:379]:
```suggestion
    m : int, required
        Index of the rectangular antenna array element in the x direction.
    n : int, required
        Index of the rectangular antenna array element in the y direction.
    theta_name : str, optional
        Postprocessing variable name in HFSS to use for the
        theta component of the phase angle expression. The default is ``"theta_scan"``.
    phi_name : str, optional
        Postprocessing variable name in HFSS to use to generate
        the phi component of the phase angle expression. The default is ``"phi_scan"``
```
We usually only put "optional," with the assumption that a parameter is required if "optional" isn't present. I'm not sure we use "require" anywhere elese.

File: pyaedt/application/analysis_hf.py [hunk 2]
```diff
@@ -360,3 +360,64 @@ def export_touchstone(
...
+
+    Returns
+    -------
+    str
+        The phase angle expression for the (m,n) source of
+        the (m,n) antenna array element.
+
+    """
+    # px is the term for the phase variation in the x-direction
```
Comment (PipKat) [inline pyaedt/application/analysis_hf.py:388]:
```suggestion
    # px is the term for the phase variation in the x-direction.
```

File: pyaedt/application/analysis_hf.py [hunk 3]
```diff
@@ -360,3 +360,64 @@ def export_touchstone(
...
+    # py is the term for the phase variation in the y-direction.
```
Comment (PipKat) [inline pyaedt/application/analysis_hf.py:389]:
```suggestion
    # px is the term for the phase variation in the x direction.
    # py is the term for the phase variation in the y direction.
```

File: pyaedt/generic/plot.py [hunk 4]
```diff
@@ -38,13 +38,37 @@
...
+        }
+
     except ImportError:
         warnings.warn(
             "The Matplotlib module is required to run some functionalities of PostProcess.\n"
             "Install with \n\npip install matplotlib\n\nRequires CPython."
         )
     except Exception:
-        pass
+        warnings.warn("Unknown error while attempting to import matplotlib.")
```
Comment (PipKat) [inline pyaedt/generic/plot.py:54]:
```suggestion
        warnings.warn("Unknown error occurred while attempting to import Matplotlib.")
```
String should have a verb (as well as concluding punctuation).

File: pyaedt/generic/plot.py [hunk 5]
```diff
@@ -293,8 +318,9 @@ def _parse_streamline(filepath):
 
 
 @pyaedt_function_handler()
+@update_plot_settings
 def plot_polar_chart(
-    plot_data, size=(2000, 1000), show_legend=True, xlabel="", ylabel="", title="", snapshot_path=None
+    plot_data, size=(2000, 1000), show_legend=True, xlabel="", ylabel="", title="", snapshot_path=None, show=True
 ):
     """Create a matplotlib polar plot based on a list of data.
```
Comment (PipKat) [inline pyaedt/generic/plot.py:325]:
```suggestion
    """Create a Matplotlib polar plot based on a list of data.
```

File: pyaedt/generic/plot.py [hunk 6]
```diff
@@ -315,6 +341,14 @@ def plot_polar_chart(
         Plot Title label.
```
Comment (PipKat) [inline pyaedt/generic/plot.py:341]:
```suggestion
        Plot title label.
```

File: pyaedt/generic/plot.py [hunk 7]
```diff
@@ -315,6 +341,14 @@ def plot_polar_chart(
...
     snapshot_path : str
         Full path to image file if a snapshot is needed.
```
Comment (PipKat) [inline pyaedt/generic/plot.py:343]:
```suggestion
        Full path to the image file if a snapshot is needed.
```

File: pyaedt/generic/plot.py [hunk 8]
```diff
@@ -315,6 +341,14 @@ def plot_polar_chart(
...
+    show : Bool
+        Set to ``True`` if the figure should be rendered. Default is ``True``, otherwise
+        figure is not drawn.
```
Comment (PipKat) [inline pyaedt/generic/plot.py:346]:
```suggestion
    show : bool, optional
        Whether to render the figure. The default is ``True``. If ``False``, the
        figure is not drawn.
```

File: pyaedt/generic/plot.py [hunk 9]
```diff
@@ -315,6 +341,14 @@ def plot_polar_chart(
...
+
+    Returns
+    -------
+    :class:`matplotlib.pyplot.Figure`
+        Matplotlib fig object.
```
Comment (PipKat) [inline pyaedt/generic/plot.py:351]:
```suggestion
        Matplotlib figure object.
```

File: pyaedt/generic/plot.py [hunk 10]
```diff
@@ -344,12 +378,13 @@ def plot_polar_chart(
...
-    else:
+    if show:
         fig.show()
     return fig
 
 
 @pyaedt_function_handler()
+@update_plot_settings
 def plot_3d_chart(plot_data, size=(2000, 1000), xlabel="", ylabel="", title="", snapshot_path=None):
     """Create a matplotlib 3D plot based on a list of data.
```
Comment (PipKat) [inline pyaedt/generic/plot.py:389]:
```suggestion
    """Create a Matplotlib 3D plot based on a list of data.
```

File: pyaedt/generic/plot.py [hunk 11]
```diff
@@ -371,7 +406,7 @@ def plot_3d_chart(plot_data, size=(2000, 1000), xlabel="", ylabel="", title="",
 
     Returns
     -------
-    :class:`matplotlib.plt`
+    :class:`matplotlib.pyplot.Figure`
         Matplotlib fig object.
```
Comment (PipKat) [inline pyaedt/generic/plot.py:410]:
```suggestion
        Matplotlib figure object.
```

File: pyaedt/generic/plot.py [hunk 12]
```diff
@@ -403,9 +438,9 @@ def plot_3d_chart(plot_data, size=(2000, 1000), xlabel="", ylabel="", title="",
 
 
 @pyaedt_function_handler()
+@update_plot_settings
 def plot_2d_chart(plot_data, size=(2000, 1000), show_legend=True, xlabel="", ylabel="", title="", snapshot_path=None):
     """Create a matplotlib plot based on a list of data.
```
Comment (PipKat) [inline pyaedt/generic/plot.py:443]:
```suggestion
    """Create a Matplotlib plot based on a list of data.
```

File: pyaedt/generic/plot.py [hunk 13]
```diff
@@ -427,7 +462,7 @@ def plot_2d_chart(plot_data, size=(2000, 1000), show_legend=True, xlabel="", yla
 
     Returns
     -------
-    :class:`matplotlib.plt`
+    :class:`matplotlib.pyplot.Figure`
         Matplotlib fig object.
```
Comment (PipKat) [inline pyaedt/generic/plot.py:466]:
```suggestion
        Matplotlib figure object.
```

File: pyaedt/generic/plot.py [hunk 14]
```diff
@@ -569,13 +605,16 @@ def plot_matplotlib(
...
+    return fig
 
 
 @pyaedt_function_handler()
-def plot_contour(qty_to_plot, x, y, size=(2000, 1600), xlabel="", ylabel="", title="", levels=64, snapshot_path=None):
+@update_plot_settings
+def plot_contour(
+    qty_to_plot, x, y, size=(2000, 1600), xlabel="", ylabel="", title="", levels=64, snapshot_path=None, show=True
+):
     """Create a matplotlib contour plot.
```
Comment (PipKat) [inline pyaedt/generic/plot.py:618]:
```suggestion
    """Create a Matplotlib contour plot.
```

File: pyaedt/generic/plot.py [hunk 15]
```diff
@@ -598,10 +637,13 @@ def plot_contour(qty_to_plot, x, y, size=(2000, 1600), xlabel="", ylabel="", tit
         Color map levels. Default is `64`.
     snapshot_path : str, optional
         Full path to image to save. Default is None.
+    show : Bool, optional
+        Set to ``True`` if the figure should be rendered.
+        Default is ``True``.
```
Comment (PipKat) [inline pyaedt/generic/plot.py:642]:
```suggestion
        Color map levels. The default is ``64``.
    snapshot_path : str, optional
        Full path to save the image save. The default is ``None``.
    show : bool, optional
        Whether to render the figure. The default is ``True``. If
        ``False``, the image is not drawn.
```
The edits were to make the descriptions consistent and follow stye guidelines. I feel like we shouldn't switch from figure to image unless there is some reason to. The key is to understanding is the use of consistent terms.

File: pyaedt/generic/plot.py [hunk 16]
```diff
@@ -598,10 +637,13 @@ def plot_contour(qty_to_plot, x, y, size=(2000, 1600), xlabel="", ylabel="", tit
...
 
     Returns
     -------
-    :class:`matplotlib.plt`
+    :class:`matplotlib.pyplot.Figure`
         Matplotlib fig object.
```
Comment (PipKat) [inline pyaedt/generic/plot.py:647]:
```suggestion
        Matplotlib figure object.
```

File: pyaedt/hfss.py [hunk 17]
```diff
@@ -5237,13 +5237,14 @@ def set_differential_pair(
 
     @pyaedt_function_handler(array_name="name", json_file="input_data")
     def add_3d_component_array_from_json(self, input_data, name=None):
-        """Add or edit a 3D component array from a JSON file or TOML file.
+        """Add or edit a 3D component array from a JSON file, TOML file or dict.
```
Comment (PipKat) [inline pyaedt/hfss.py:5240]:
```suggestion
        """Add or edit a 3D component array from a JSON file, TOML file, or dictionary.
```

File: pyaedt/hfss.py [hunk 18]
```diff
@@ -5237,13 +5237,14 @@ def set_differential_pair(
...
         The 3D component is placed in the layout if it is not present.
 
         Parameters
         ----------
         input_data : str, dict
-            Full path to either the JSON file or dictionary containing the array information.
+            Full path to either the JSON file, TOML file or the dictionary
```
Comment (PipKat) [inline pyaedt/hfss.py:5246]:
```suggestion
            Full path to either the JSON file, TOML file, or the dictionary
```

File: pyaedt/hfss.py [hunk 19]
```diff
@@ -5417,8 +5418,11 @@ def get_antenna_ffd_solution_data(
         sphere=None,
         variations=None,
         overwrite=True,
+        link_to_hfss=True,
     ):
-        """Export antennas parameters to Far Field Data (FFD) files and return the ``FfdSolutionDataExporter`` object.
+        """Export antennas parameters to Far Field Data (FFD) files and return an
```
Comment (PipKat) [inline pyaedt/hfss.py:5423]:
```suggestion
        """Export the antenna's parameters to Far Field Data (FFD) files and return an
```

File: pyaedt/hfss.py [hunk 20]
```diff
@@ -5417,8 +5418,11 @@ def get_antenna_ffd_solution_data(
...
+        instance of
```
Comment (PipKat) [inline pyaedt/hfss.py:5424]:
```suggestion
        instance of the
```

File: pyaedt/hfss.py [hunk 21]
```diff
@@ -5435,12 +5439,19 @@ def get_antenna_ffd_solution_data(
             Variation dictionary.
         overwrite : bool, optional
             Whether to overwrite FFD files. The default is ``True``.
+        link_to_hfss : bool, optional
+            If this is set to ``False`` then return an instance of
+            :class:`pyaedt.modules.solutions.FfdSolutionData` which is independent from the
+            running HFSS instance. The default is ``True`` which returns an instance of
+            :class:`pyaedt.modules.solutions.FfdSolutionDataExporter` which requires a connection
+            to an instance of the :class:`Hfss`` class.
```
Comment (PipKat) [inline pyaedt/hfss.py:5447]:
```suggestion
        link_to_hfss : bool, optional
            Whether to return an instance of the
            :class:`pyaedt.modules.solutions.FfdSolutionDataExporter` class,
            which requires a connection to an instance of the :class:`Hfss` class.
            The default is `` True``. If ``False``, returns an instance of
            :class:`pyaedt.modules.solutions.FfdSolutionData` class, which is
            independent from the running HFSS instance.
```

File: pyaedt/modules/solutions.py [hunk 22]
```diff
@@ -1088,18 +1091,21 @@ def ifft_to_file(
 
 
 class FfdSolutionData(object):
-    """Contains information from the far field solution data.
+    """Antenna array far-field data class.
```
Comment (PipKat) [inline pyaedt/modules/solutions.py:1094]:
```suggestion
    """Provides antenna array far-field data.
```
A description of a class should start with a verb that ends in "s" or "es"

File: pyaedt/modules/solutions.py [hunk 23]
```diff
@@ -1088,18 +1091,21 @@ def ifft_to_file(
...
 
-    Load far field data from the element pattern files.
+    Read embedded element patterns generated in HFSS and provide the Python interface
```
Comment (PipKat) [inline pyaedt/modules/solutions.py:1096]: The word "provide" seems odd. Perhaps replace with "use" instead?

File: pyaedt/modules/solutions.py [hunk 24]
```diff
@@ -1088,18 +1091,21 @@ def ifft_to_file(
...
+    to plot and analyze the array far-field data.
 
     Parameters
     ----------
     eep_files : list or str
-        List of element pattern files for each frequency.
-        If the input is string, it is assumed to be a single frequency.
+        List of embedded element pattern files for each frequency.
+        If data is only provided for a single frequency, then a string may be passed
```
Comment (PipKat) [inline pyaedt/modules/solutions.py:1103]:
```suggestion
        If data is only provided for a single frequency, then a string can be passed
```
May should be reserved for granting permission. Use "can" or "might" instead.

File: pyaedt/modules/solutions.py [hunk 25]
```diff
@@ -1088,18 +1091,21 @@ def ifft_to_file(
...
+        instead of a 1-element list.
```
Comment (PipKat) [inline pyaedt/modules/solutions.py:1104]:
```suggestion
        instead of a one-element list.
```

File: pyaedt/modules/solutions.py [hunk 26]
```diff
@@ -1088,18 +1091,21 @@ def ifft_to_file(
...
     frequencies : list, str, int, or float
         List of frequencies.
-        If the input is not a list, it is assumed to be a single frequency.
+        If data is only available for a single frequency, then a float or int may be passed
```
Comment (PipKat) [inline pyaedt/modules/solutions.py:1107]:
```suggestion
        If data is only available for a single frequency, then a float or integer may be passed
```

File: pyaedt/modules/solutions.py [hunk 27]
```diff
@@ -1088,18 +1091,21 @@ def ifft_to_file(
...
+        instead of a 1-element list.
```
Comment (PipKat) [inline pyaedt/modules/solutions.py:1108]:
```suggestion
        instead of a one-element list.
```

File: pyaedt/modules/solutions.py [hunk 28]
```diff
@@ -1546,6 +1555,9 @@ def plot_farfield_contour(
             Phi scan angle in degrees. The default is ``0``.
         theta : float, int, optional
             Theta scan angle in degrees. The default is ``0``.
+        size : tuple, optional
+            Image size in pixel (width, height). Default is ``None`` in which case resolution
+            is determined automatically.
```
Comment (PipKat) [inline pyaedt/modules/solutions.py:1560]:
```suggestion
            Image size in pixel (width, height). The default is ``None``, in which case resolution
            is determined automatically.
```

File: pyaedt/modules/solutions.py [hunk 29]
```diff
@@ -1559,25 +1571,31 @@ def plot_farfield_contour(
         show : bool, optional
             Whether to show the plot. The default is ``True``. If ``False``, the Matplotlib
             instance of the plot is shown.
+        polar : bool, optional
+            Generate the plot in polar coordinates. The default is ``True``. If ``False``, the plot
+            will be rectangular.
+        max_theta : float or int, optional
+            Maxmum theta angle for plotting. The default is ``180`` which plots the far-field for
+            all angles. Setting ``max_theta`` to 90 will limit the displayed data to the upper
+            hemisphere, that is (0 < theta < 90).
```
Comment (PipKat) [inline pyaedt/modules/solutions.py:1580]:
```suggestion
            Maxmum theta angle for plotting. The default is ``180``, which plots the far-field data for
            all angles. Setting ``max_theta`` to 90 limits the displayed data to the upper
            hemisphere, that is (0 < theta < 90).
```

File: pyaedt/modules/solutions.py [hunk 30]
```diff
@@ -2420,7 +2465,19 @@ def _rotation_to_euler_angles(R):
 
 
 class FfdSolutionDataExporter(FfdSolutionData):
-    """Export far field solution data.
+    """Class to enable export of embedded element pattern data from HFSS.
```
Comment (PipKat) [inline pyaedt/modules/solutions.py:2468]:
```suggestion
    """EnableS export of embedded element pattern data from HFSS.
```
We don't need to indicate that this is a class as its clear here and in the documentation.

File: pyaedt/modules/solutions.py [hunk 31]
```diff
@@ -2420,7 +2465,19 @@ def _rotation_to_euler_angles(R):
...
+
+    An instance of this class is returned from the method
```
Comment (PipKat) [inline pyaedt/modules/solutions.py:2470]:
```suggestion
    An instance of this class is returned from the
```

File: pyaedt/modules/solutions.py [hunk 32]
```diff
@@ -2420,7 +2465,19 @@ def _rotation_to_euler_angles(R):
...
+    :meth:`pyaedt.Hfss.get_antenna_ffd_solution_data`. This method allows creation of
```
Comment (PipKat) [inline pyaedt/modules/solutions.py:2471]:
```suggestion
    :meth:`pyaedt.Hfss.get_antenna_ffd_solution_data` method. This method allows creation of
```
You should put the object (method) after the tagged code entity, not before.

File: pyaedt/modules/solutions.py [hunk 33]
```diff
@@ -2420,7 +2465,19 @@ def _rotation_to_euler_angles(R):
...
+    the embedded
+    element pattern (eep) files for an antenna array that has been solved in HFSS. The
```
Comment (PipKat) [inline pyaedt/modules/solutions.py:2473]:
```suggestion
    element pattern (EEP) files for an antenna array that have been solved in HFSS. The
```

File: pyaedt/modules/solutions.py [hunk 34]
```diff
@@ -2420,7 +2465,19 @@ def _rotation_to_euler_angles(R):
...
+    ``frequencies`` and ``eep_files`` properties can then be passed as arguments to
+    instantiate an instance of :class:`pyaedt.modules.solutions.FfdSolutionData` for
+    subsequent analysis and post-processing of the array data.
```
Comment (PipKat) [inline pyaedt/modules/solutions.py:2476]:
```suggestion
    instantiate an instance of the :class:`pyaedt.modules.solutions.FfdSolutionData` class for
    subsequent analysis and postprocessing of the array data.
```

File: pyaedt/modules/solutions.py [hunk 35]
```diff
@@ -2420,7 +2465,19 @@ def _rotation_to_euler_angles(R):
...
+
+    Note that this class is derived from :class:`FfdSolutionData` and can be used directly for
+    far-field postprocessing and array analysis, but remains a property of the
+    :class:`pyaedt.Hfss` application.
```
Comment (PipKat) [inline pyaedt/modules/solutions.py:2480]:
```suggestion
    Note that this class is derived from the :class:`FfdSolutionData` class and can be used directly for
    far-field postprocessing and array analysis, but it remains a property of the
    :class:`pyaedt.Hfss` application.
```

**Review by SMoraisAnsys — COMMENTED** [review]
LGTM
@dcrawforAtAnsys Thanks for this improvement!
@Samuelopez-ansys Thanks for helping with debugging. What was the issue ?

**Comment by Samuelopez-ansys** [comment]
@SMoraisAnsys After debugging, I found that with the new decorator, the plot is opened in the unit test, but until the user do not close it, the script/test is alive forever. We need to have two options, check a matplotlib option to allow going ahead with the script when show=True, or like I did, do not test the show=True case. We can discuss it later.

@dcrawforAtAnsys

File: pyaedt/generic/plot.py (hunk 15 above)
Comment (dcrawforAtAnsys) [reply pyaedt/generic/plot.py:642]: @PipKat  in this context "figure" refers to a Python object while "image" refers to the rendering of that figure.

File: pyaedt/hfss.py (hunk 19 above)
Comment (dcrawforAtAnsys) [reply pyaedt/hfss.py:5423]: @PipKat In this context "antenna parameter" is correct or the plural. This is clear to engineers working with antennas.

File: pyaedt/modules/solutions.py (hunk 23 above)
Comment (dcrawforAtAnsys) [inline pyaedt/modules/solutions.py:1096]:
```suggestion
    Read embedded element patterns generated in HFSS and return the Python interface
```

**Review by Devin-Crawford — COMMENTED** [review]
Suggestions were added.

**Review by Samuelopez-ansys — APPROVED** [review]

**Comment by Samuelopez-ansys** [comment]
@dcrawforAtAnsys Can we merge?

**Pull request merged**